### DISCUSSION TITLE
[Chapter 08] 다대다 관계

### DISCUSSION THREAD

**Discussion by JoisFe** [cowaine/spring-boot-for-msa#128] [Chapter 08. Data Persistence]
책에서 다대대 관계 예시를 보여주기 위해
`reservations` 테이블과 `payments` 테이블의 다대다 관계를 보여주고 있습니다.

중간에 매핑 테이블로 `reservations_payments_map` 이라는 테이블을 두고
실제 DB 상 테이블에서는 `reservations`와 `reservations_payments_map`이 일대다 관계
`payments`와 `reservations_payments_map`이 일대다 관계를 두고 있습니다.
그런데 책에서는 매핑 테이블인 `reservations_payments_map` 엔티티를 만들지 않고
ReservationEntity와 PaymentEntity 간의 `@ManyToMany`를 이용하여
``` java
@JoinTable(name="reservations_payments_map" ~~) 
```
을 두어 다대다 관계를 맺고 있습니다.

여기서 질문이 책 방법 말고
그냥 `reservations_payments_map` 엔티티 객체를 두고 `@ManyToMany`를 사용하지 않고
`@OneToMany`로 ReservationEntity와 reservations_payments_map 엔티티와 일대다 관계를
또한 PaymentEntity와  reservations_payments_map 엔티티와 일대다 관계를 맺어주어도 될 것 같은데
이 방법도 괜찮은지, 그리고 괜찮다면 두 방법간의 차이는 무엇인지 궁금합니다.

**Comment by corock** [comment]
`@ManyToMany` 애너테이션은 안 괜찮은 방법으로 보입니다. 안 괜찮은 방법을 넘어서 사용을 막아야 할 것 같네요.
일단 관계형 데이터베이스에서부터 테이블 2개로 다대다 관계를 표현할 수 없고, 매핑 테이블이라는 녀석의 확장 가능성을 생각해 봐야 합니다.

`reservations_payments_map` 테이블을 예로 들면, 매핑에 필요한 필드 외에도 이 테이블에서 부가 정보가 언제든지 추가될 수 있습니다.
이는 JPA 를 통해서 예상하지 못한 쿼리를 실행할 수 있는 여지를 남기고, 성능 저하의 원인이 될 수도 있지 않을까 싶습니다.
제가 접하는 현장에서조차도(엔티티 매핑해 놓은 걸 보면 현기증이 나지만) `@ManyToMany` 은 하나도 없습니다.

_cf._ 찾아보니 다대다 테이블 간 연결을 지어주는 매핑 테이블을 연결 테이블(_Junction Table_)이라고 하네요.
